server/app/query/view/[id]/components.tsx

```diff
@@ -65,12 +66,37 @@ function secondsToTime(e: number) {
 
 export function RunTimePill({
   status,
-  runTime,
+  startTime,
+  endTime,
 }: {
   status: Status;
-  runTime: number | null;
+  startTime: number | null;
+  endTime: number | null;
 }) {
+  const [runTime, setRunTime] = useState<number | null>(null);
   const runTimeStr = runTime ? secondsToTime(runTime) : "N/A";
+  const intervalId = useRef<ReturnType<typeof setTimeout> | null>(null);
+
+  useEffect(() => {
+    if (intervalId.current !== null) {
+      // any time startTime or endTime change, we remove the old setInterval
+      // which runs the timer. if a new one is needed, it's created.
+      clearTimeout(intervalId.current);
+    }
+    if (startTime === null) {
+      setRunTime(null);
+    } else {
+      if (endTime !== null && startTime !== null) {
+        setRunTime(endTime - startTime);
+      } else {
+        let newIntervalId = setInterval(() => {
+          setRunTime(Date.now() / 1000 - startTime);
```

Comment:
can't you end up here with `startTime == null`?

Reply:
should be covered above on line 86

Reply:
ah, is there a reason for checking `startTime` again on the line 89?

Reply:
oh, nope! just a vestige of a refactor.